### PR TITLE
ENH: mypy: get specific about ignoring missing imports

### LINKED ISSUE #11686: MAINT: mypy imports of Cython "modules"
For kicks following the merge of #11678, I checked on a module I'm familiar with by disabling the ignores and I see:

```
scipy/spatial/_spherical_voronoi.py:17: error: Module 'scipy.spatial' has no attribute '_voronoi'  [attr-defined]
scipy/spatial/_spherical_voronoi.py:18: error: Module 'scipy.spatial' has no attribute 'cKDTree'; maybe "KDTree"?  [attr-defined]
```

That's because those are Cython extension objects rather than pure Python modules.

Can be suppressed in this way with a simple inline comment--would we want to do this more broadly throughout code base?

```diff
diff --git a/scipy/spatial/_spherical_voronoi.py b/scipy/spatial/_spherical_voronoi.py
index f14b894aa..f7f86e540 100644
--- a/scipy/spatial/_spherical_voronoi.py
+++ b/scipy/spatial/_spherical_voronoi.py
@@ -14,8 +14,8 @@ Spherical Voronoi Code
 import warnings
 import numpy as np
 import scipy
-from . import _voronoi
-from scipy.spatial import cKDTree
+from . import _voronoi # type: ignore
+from scipy.spatial import cKDTree # type: ignore
 
 __all__ = ['SphericalVoronoi']
```

### PR DESCRIPTION
**Pull request by person142**
#### Reference issue
Closes https://github.com/scipy/scipy/issues/11686.

#### What does this implement/fix?
In the first iteration of the ratchet, we just ignored all missing
imports. This:

- Adds a `mypy_requirements.txt` file that will add the latest version
  of the NumPy types and stops ignoring NumPy imports
- Ignores all other third party modules
- Ignores extension modules (for now) since mypy can't find those
  without stubs

#### Additional information
Note that I'm leaving mypy in a dirty state-mostly because it appears
to have found some things worth looking into:

```
scipy/linalg/setup.py:121: error: Cannot find implementation or library stub for module named 'linalg_version'  [import]
scipy/linalg/setup.py:121: note: See https://mypy.readthedocs.io/en/latest/running_mypy.html#missing-imports
scipy/sparse/linalg/_norm.py:7: error: Module 'numpy.core' has no attribute 'Inf'  [attr-defined]
scipy/sparse/linalg/_norm.py:7: error: Module 'numpy.core' has no attribute 'sqrt'  [attr-defined]
scipy/sparse/linalg/_norm.py:7: error: Module 'numpy.core' has no attribute 'abs'  [attr-defined]
scipy/fftpack/tests/test_pseudo_diffs.py:325: error: List item 0 has incompatible type "Type[complex64]"; expected "Type[floating]"  [list-item]
scipy/fftpack/tests/test_pseudo_diffs.py:325: error: List item 1 has incompatible type "Type[complex128]"; expected "Type[floating]"  [list-item]
scipy/fftpack/tests/test_basic.py:758: error: List item 0 has incompatible type "Type[complex64]"; expected "Type[floating]"  [list-item]
scipy/fftpack/tests/test_basic.py:758: error: List item 1 has incompatible type "Type[complex128]"; expected "Type[floating]"  [list-item]
scipy/io/matlab/tests/test_mio.py:140: error: Argument 2 to "zeros" has incompatible type "List[Tuple[str, Type[object]]]"; expected "Union[dtype, None, type, str, Tuple[Any, int], Tuple[Any, Union[int, Sequence[int]]], List[Union[Tuple[Union[str, Tuple[str, str]], Any], Tuple[Union[str, Tuple[str, str]], Any, Union[int, Sequence[int]]]]], Dict[str, Union[Sequence[str], Sequence[Any], Sequence[int], Sequence[Union[bytes, str, None]], int]], Dict[str, Tuple[Any, int]], Tuple[Any, Any]]"  [arg-type]
scipy/io/matlab/tests/test_mio.py:140: note: "List" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
scipy/io/matlab/tests/test_mio.py:140: note: Consider using "Sequence" instead, which is covariant
scipy/linalg/tests/test_decomp_cossin.py:12: error: Unsupported operand types for + ("List[Type[floating]]" and "List[Type[complexfloating]]")  [operator]
Found 10 errors in 6 files (checked 626 source files)
```

One of the things is https://github.com/scipy/scipy/pull/11687, this one

```
scipy/linalg/setup.py:121: error: Cannot find implementation or library stub for module named 'linalg_version'  [import]
```

appears to be because we should be doing a relative import here:

https://github.com/scipy/scipy/blob/master/scipy/linalg/setup.py#L121

### CODE REVIEW

File: mypy_requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+mypy==0.770
+git+git://github.com/numpy/numpy-stubs.git@master
```
Comment (person142) [inline mypy_requirements.txt]: In these very early stages it seems fair to pull from the master branch.

Comment (rgommers) [reply mypy_requirements.txt]: +1

**Comment by rgommers** [comment]
> Note that I'm leaving mypy in a dirty state-mostly because it appears
> to have found some things worth looking into:

That seems fine to me, can be fixed in follow-ups.

Merging, thanks @person142

**Comment by rgommers** [comment]
CI failure was an often occurring flake on `differential_evolution`. There doesn't seem to be an open issue though, @andyfaff does that failure look familiar to you?

**Comment by rgommers** [comment]
@person142 I mentioned this MyPy push and the status of `numpy-stubs` in the NumPy community call last night. There was quite a bit of interest in getting typing for `ndarray` and `array_like` (see https://github.com/numpy/numpy-stubs/issues/37) right and released. Step 1 of the plan in the README of https://github.com/numpy/numpy-stubs basically. Scikit-image, Napari and another imaging library already have a use for that.

**Comment by andyfaff** [comment]
@rgommers, I've not seen that failure before. The solver is seeded, so I would expect the output of the stochastic part to be the same. It could be due to SLSQP giving slightly different answers on a different platform?

**Pull request merged**